File: vertx-elasticsearch/src/main/java/fr/myprysm/vertx/elasticsearch/action/search/aggregations/Aggregation.java
package fr.myprysm.vertx.elasticsearch.action.search.aggregations;

import io.vertx.codegen.annotations.DataObject;
import io.vertx.core.json.JsonObject;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Accessors(chain = true)
@DataObject(generateConverter = true)
public class Aggregation {
    private String type;
    private String name;
    private JsonObject metaData;
    private JsonObject data;

    public Aggregation(Aggregation other) {
        type = other.type;
        name = other.name;
        metaData = other.metaData;
        data = other.data;
    }

    /**
     * Build a new <code>Aggregation</code> from a <code>JsonObject</code>.
     *
     * @param json the <code>JsonObject</code>
     */
    public Aggregation(JsonObject json) {
        AggregationConverter.fromJson(json, this);
    }

    /**
     * Transforms the <code>Aggregation</code> into a <code>JsonObject</code>.
     *
     * @return the <code>JsonObject</code>
     */
    public JsonObject toJson() {
        JsonObject json = new JsonObject();
        AggregationConverter.toJson(this, json);
        return json;
    }
}

File: vertx-elasticsearch/src/main/kotlin/fr/myprysm/vertx/elasticsearch/kotlin/action/search/aggregations/Aggregation.kt
package fr.myprysm.vertx.elasticsearch.kotlin.action.search.aggregations

import fr.myprysm.vertx.elasticsearch.action.search.aggregations.Aggregation

fun Aggregation(
        data: io.vertx.core.json.JsonObject? = null,
        metaData: io.vertx.core.json.JsonObject? = null,
        name: String? = null,
        type: String? = null): Aggregation = fr.myprysm.vertx.elasticsearch.action.search.aggregations.Aggregation().apply {

    if (data != null) {
        this.setData(data)
    }
    if (metaData != null) {
        this.setMetaData(metaData)
    }
    if (name != null) {
        this.setName(name)
    }
    if (type != null) {
        this.setType(type)
    }
}

